Blog: Read support for original files of Hive transactional tables

harmandeeps opened the pull request.

findepi commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+
+In Presto 331, read support for Hive transactional tables was introduced. It works well if a user creates a new Hive transactional table and reads it from Presto. However,  if an
+existing table is converted to a Hive transactional table, Presto would fail to read data from such a table because  read support for original files was missing. Original files are
+those files in a Hive transactional table that existed before the table was converted into a Hive transactional table and until version 340, Presto expected all files in a Hive
```
340 -> ...

(not merged yet, sorry for the delay)

this applies in other places too

harmandeeps replied:
okay, I will modify the version when finalised.

findepi replied:
341 has all the changes, am I right?

findepi replied:
341 has all the changes already merged, am I right?

harmandeeps replied:
I can see the commit mentioned in 340 release notes, sorry I am confused, shouldn't it be 340 then?

findepi replied:
You're right. I mixed this up with a different change.

findepi commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+transactional table to be in Hive ACID format. Users would have to perform a Major Compaction to convert original files into ACID files (i.e. base files) in such tables. This is
+not always possible as the original flat table (table in non-ACID format) could be huge and converting all the existing data into ACID format would be very expensive.
+
+This blog is an extension of the blog [Hive ACID and transactional tables' support in Presto](https://prestosql.io/blog/2020/06/01/hive-acid.html). It first describes original
```
Use `{% post_url` for cross links.

(this applies below too)

> This blog is an extension of the blog

i am not native, but maybe something like
```
See [....] for more information about Hive ORC AID and transactional tables structure.
```

mosabua replied:
The existing text is good imho.. but yes .. use the post url stuff

harmandeeps replied:
fixed.

findepi commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+
+### Original transaction ID
+
+An original Transaction ID is the write ID when a record is first created. For original files, the original transaction ID is always the starting value i.e. 0.
```
```suggestion
An original Transaction ID is the write ID when a record is first created. For original files, the original transaction ID is always 0.
```

harmandeeps replied:
fixed.

findepi commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+
+### Bucket ID
+
+Bucket ID is retrieved from the original file name. For the original file `0000ABC_DEF` or `0000ABC_DEF_copy_G`, the bucket ID will be `DEF`.
```
Not ABC ?

harmandeeps replied:
yeah, thank you for pointing out. changed.

findepi commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+000000_0_copy_3     ->  X4 Rows
+~~~~
+
+>**Note:** As we see, additional computations are required to generate rowIDs while reading original files, therefore, read will be slower than ACID format files in the transactional
```
rowIDs -> Row ID

(also below)

harmandeeps replied:
fixed.

findepi commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+each partition. In addition to the steps mentioned in blog [Hive ACID and transactional tables' support in Presto](https://prestosql.io/blog/2020/06/01/hive-acid.html), Presto does
+the following:
+
+1. Original files, ACID subdirectories (`base`, `delta`, `delete_delta`) are figured out by listing the partition location by Hive AcidUtils Helper class.
```
> by Hive AcidUtils Helper class

are we going too low level for a blog post?

harmandeeps replied:
I added high-level changes for the better understanding. If you think this is not needed, then we can omit

findepi commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+
+# Follow up
+
+For an original file split, the current implementation may take polynomial time in the worst case to calculate global rowID by reading rowIDs from the original files’ footer. It
```
polynomial -> quadratic?

harmandeeps replied:
fixed.

findepi commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+# Acknowledgements
+
+I would also like to express my gratitude to everyone who helped me throughout developing the feature. Thank you [Shubham Tagra](https://in.linkedin.com/in/shubham-tagra-267a5838)
+for brainstorming sessions and providing continuous guidance on Presto Hive ACID. Thank you [Piotr](https://www.linkedin.com/in/piotrfindeisen/) for helping me further refine the
```
Thanks!

findepi commented:
I left some comments, but was trying not to review language or communication.
I am not native. Also, I know how it works, so I am not the right person to determine
whether the text can be easily understood.

@mosabua would you like to give this a read?
@wendigo maybe you'd be interested too?

wendigo commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+
+# What are original files?
+
+Files present in flat ORC tables have the standard ORC schema. When a flat table is converted into a transactional table, existing files are not converted into Hive ACID format.
```
Maybe instead of using "flat" table you can use non-transactional term instead?

harmandeeps replied:
fixed.

wendigo commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+---
+
+In Presto 331, read support for Hive transactional tables was introduced. It works well if a user creates a new Hive transactional table and reads it from Presto. However,  if an
+existing table is converted to a Hive transactional table, Presto would fail to read data from such a table because  read support for original files was missing. Original files are
```
redundant space: "because  read"

harmandeeps replied:
fixed.

wendigo commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+not always possible as the original flat table (table in non-ACID format) could be huge and converting all the existing data into ACID format would be very expensive.
+
+This blog is an extension of the blog [Hive ACID and transactional tables' support in Presto](https://prestosql.io/blog/2020/06/01/hive-acid.html). It first describes original
+files and then goes into details of  read support for such files that was  added in Presto 340.
```
redundant double spaces

harmandeeps replied:
fixed.

wendigo commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+not always possible as the original flat table (table in non-ACID format) could be huge and converting all the existing data into ACID format would be very expensive.
+
+This blog is an extension of the blog [Hive ACID and transactional tables' support in Presto](https://prestosql.io/blog/2020/06/01/hive-acid.html). It first describes original
+files and then goes into details of  read support for such files that was  added in Presto 340.
```
if think it should be "reading support"

mosabua replied:
nope .. read support is correct

wendigo commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+
+<!--more-->
+
+# What are original files?
```
What are the original files

harmandeeps replied:
fixed.

wendigo commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+# What are original files?
+
+Files present in flat ORC tables have the standard ORC schema. When a flat table is converted into a transactional table, existing files are not converted into Hive ACID format.
+Such files, in a transactional table which are not in Hive ACID format, are called original files. These files are named as `000000_X`, `000000_X_copy_Y`. These files don’t have ACID
```
s/which are/that are

harmandeeps replied:
fixed.

wendigo commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+
+
+
+Before Presto 340, Presto used to fail the query if it reads from a  Hive transactional table having original files.
```
redundant double space

harmandeeps replied:
fixed.

wendigo commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+## ACID Columns’ generation on original files
+
+Files in Hive ACID format have 5 ACID columns, but we need only 3 columns i.e. `originalTransactionId`, `bucketId` and `rowId` to uniquely identify a row. In this section we will
+see how these 3 columns are synthetically generated for  original files.
```
redundant double space

harmandeeps replied:
fixed.

wendigo commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+
+## ACID Columns’ generation on original files
+
+Files in Hive ACID format have 5 ACID columns, but we need only 3 columns i.e. `originalTransactionId`, `bucketId` and `rowId` to uniquely identify a row. In this section we will
```
In this section, we will

harmandeeps replied:
fixed.

wendigo commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+
+## Split generation
+
+We use a new construct `AcidInfo` to store `OriginalFiles`, `DeleteDeltaFiles` for HiveSplit. `BackgroundSplitLoader.loadPartitions` is called in an executor to create splits for
```
> We use a new construct `AcidInfo` to store `OriginalFiles`, `DeleteDeltaFiles` for HiveSplit.

We use a new class named `AcidInfo` to store `OriginalFiles` and`DeleteDeltaFiles` for `HiveSplit`.

harmandeeps replied:
fixed.

wendigo reviewed (commented):
Will comment more :) Stay tuned

mosabua commented:
General comment - 80 char line width or at least less than 120 makes reviews easier..

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+Hive achieves updates/deletes on a row in original files by synthetically generating ACID columns for original files. Presto follows the same mechanism of generating ACID columns
+synthetically as discussed later.
+
+## ACID Columns’ generation on original files
```
ACID column generation on original files

harmandeeps replied:
fixed.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+
+### Row ID
+
+Row ID is calculated by taking the summation of row counts over all the files which come before the current original file when sorted in a lexicographical order belonging to the
```
Break up this long sentence please

harmandeeps replied:
I rephrased it, please take a look.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+000000_0_copy_3     ->  X4 Rows
+~~~~
+
+>**Note:** As we see, additional computations are required to generate rowIDs while reading original files, therefore, read will be slower than ACID format files in the transactional
```
read is slower than..

harmandeeps replied:
fixed.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+>**Note:** As we see, additional computations are required to generate rowIDs while reading original files, therefore, read will be slower than ACID format files in the transactional
+table.
+
+Once we have the 3 ACID columns for a row, we can check for update/delete on it. Delete deltas, written by Hive for Original files, have rowIDs generated by following the same
```
General note .. avoid using "we" .. address the reader as "you" ..

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+[Hive ACID and transactional tables' support in Presto](https://prestosql.io/blog/2020/06/01/hive-acid.html) works with the original files too.
+
+# Changes in Presto to support reading original files
+
```
don't have two titles after each other without content in between

harmandeeps replied:
yea, I have added some content here too.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+
+Once we have the 3 ACID columns for a row, we can check for update/delete on it. Delete deltas, written by Hive for Original files, have rowIDs generated by following the same
+strategy as discussed above, hence, the same logic of filtering out deleted rows as discussed in
+[Hive ACID and transactional tables' support in Presto](https://prestosql.io/blog/2020/06/01/hive-acid.html) works with the original files too.
```
post url in link

harmandeeps replied:
fixed.

mosabua commented:
Ideally you could also supply a PR for that documentation that adds some of that information in the Hive connector docs.

harmandeeps commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,130 @@
+# What are original files?
+
+Files present in flat ORC tables have the standard ORC schema. When a flat table is converted into a transactional table, existing files are not converted into Hive ACID format.
+Such files, in a transactional table which are not in Hive ACID format, are called original files. These files are named as `000000_X`, `000000_X_copy_Y`. These files don’t have ACID
```
fixed.

harmandeeps reviewed (commented):
Thank you for the review. I have addressed the comments. Please take a look.

harmandeeps commented:
@wendigo , @findepi , @mosabua : I have addressed the comments, Please review whenever you get time :)

harmandeeps commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -42,32 +54,41 @@ columns and have differences in the schema as follows:
 
-Here is an example to calculate the global Row ID of the 3rd row of an original File `000000_0_copy_2`.
+Here is an example to calculate the global Row ID of the 3rd row of an original
+File `000000_0_copy_2`.
```
@mosabua: Skipped L99 for 80 characters hard line wrap, since markdown was rendering it as it is.

mosabua replied:
yes .. good catch..

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+excerpt_separator: <!--more-->
+---
+
+In Presto 331, read support for Hive transactional tables was introduced. It
```
Maybe link to 331 release notes?

harmandeeps replied:
done.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+---
+
+In Presto 331, read support for Hive transactional tables was introduced. It
+works well if a user creates a new Hive transactional table and reads it from
```
works well,  ... table,

harmandeeps replied:
done.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+table, Presto would fail to read data from such a table because read support for
+original files was missing. Original files are those files in a Hive
+transactional table that existed before the table was converted into a Hive
+transactional table and until version 340, Presto expected all files in a Hive
```
Maybe ..

transactional table. Until version 340..

harmandeeps replied:
done.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+transactional table that existed before the table was converted into a Hive
+transactional table and until version 340, Presto expected all files in a Hive
+transactional table to be in Hive ACID format. Users would have to perform a
+Major Compaction to convert original files into ACID files (i.e. base files) in
```
why is this in upper case?

harmandeeps replied:
changed.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+Major Compaction to convert original files into ACID files (i.e. base files) in
+such tables. This is not always possible as the original flat table (table in
+non-ACID format) could be huge and converting all the existing data into ACID
+format would be very expensive.
```
can be very expensive

harmandeeps replied:
done.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+format would be very expensive.
+
+This blog is an extension of the blog [Hive ACID and transactional tables'
+support in Presto] {% post_url 2020-06-01-hive-acid %}. It first describes
```
remove psace between ] and { and add () around link I think.. check in other blog posts

harmandeeps replied:
fixed.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+000000_0_copy_3     ->  X4 Rows
+~~~~
+
+>**Note:** As we see, additional computations are required to generate row IDs
```
Does this render as desired?

harmandeeps replied:
![image](https://user-images.githubusercontent.com/4463893/93222445-9675b500-f78c-11ea-8e0b-b9c2fe822a4f.png)

yes, it is getting rendered as desire, attached the picture.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+following the same
+strategy as discussed above, hence, the same logic of filtering out deleted rows
+as discussed in [Hive ACID and transactional tables' support in Presto
+] {% post_url 2020-06-01-hive-acid %} works with the original files too.
```
no space between ] and {} .. and I think it needs to wrap in () for the link.. check earlier note

harmandeeps replied:
fixed.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+`, `DeleteDeltaFiles` for HiveSplit. `BackgroundSplitLoader.loadPartitions
+` is called in an executor to create splits for each partition. In addition
+to the steps mentioned in blog [Hive ACID and transactional tables' support in
+Presto] {% post_url 2020-06-01-hive-acid %}, Presto does the following:
```
link syntax

harmandeeps replied:
fixed.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+4. `AcidInfo.Builder` keeps a map
+`AcidInfo.Builder.bucketIdToOriginalFileInfoMap` of bucket ID to the list of
+original files belonging to the same bucket.
+5. Hive splits are created for each original file, base and delta directories.
```
Maybe break this up ... three nested lists seems a bit much

harmandeeps replied:
I have added them in separate lines to reduce the nesting.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+Hive splits generated during the split generation phase make their way to worker
+nodes where `OrcPageSourceFactory` is used to create `PageSource` for
+`TableScan` operator. In addition to the steps mentioned in blog [Hive ACID
+and transactional tables' support in Presto] {% post_url 2020-06-01-hive-acid %}
```
link syntax

harmandeeps replied:
fixed.

mosabua commented on _posts/2020-08-05-hive-acid-original-files.md:
```diff
@@ -0,0 +1,182 @@
+`TableScan` operator. In addition to the steps mentioned in blog [Hive ACID
+and transactional tables' support in Presto] {% post_url 2020-06-01-hive-acid %}
+, Presto does the following:
+1. `OrcDeletedRows` is created for `delete_delta` locations, if any.
```
empty line before list

harmandeeps replied:
fixed.

mosabua reviewed (changes requested):
Looking really good. Just a few minor fixes necessary. You need to ensure it renders correctly as well (esp links). Check the CI build or build locally and reach out if you need help

harmandeeps commented on _posts/2020-09-15-hive-acid-original-files.md:
```diff
@@ -0,0 +1,200 @@
+`000000_X`, `000000_X_copy_Y`. These files don’t have ACID columns and have
+differences in the schema as follows:
+
+Table Schema
```
@mosabua : I have removed the table, as it was not getting rendered correctly because each cell contained complex struct. Now, it renders perfectly.

![image](https://user-images.githubusercontent.com/4463893/93223214-71357680-f78d-11ea-9aaa-e43f3d61eef1.png)

harmandeeps commented:
> Looking really good. Just a few minor fixes necessary. You need to ensure it renders correctly as well (esp links). Check the CI build or build locally and reach out if you need help

@mosabua : Thank you for the prompt review. I have addressed your comments and verified the doc by rendering (following the steps mentioned in github documentation). Attaching a part of the blog picture for reference. Please take a look.

![image](https://user-images.githubusercontent.com/4463893/93223306-8ad6be00-f78d-11ea-9f8a-e73884350e9e.png)

mosabua commented:
Looks great now!

electrum commented:
@harmandeeps Sorry for the all the back and forth on this. Can you update the date to today and squash the commits, then I'll merge it?

harmandeeps commented:
> Looks great now!

@mosabua : Thank you for the valuable inputs.

> @harmandeeps Sorry for the all the back and forth on this. Can you update the date to today and squash the commits, then I'll merge it?

@electrum : I have changed the date to today and squashed the committees. Please have a look, Thank you.

electrum commented:
Thanks!
